File: estagio/conting/conting/select.c
/*----------------------------select.c------------------------------*/

/*        Funcoes associadas a selecao de contingencias
--------------------------------------------------  ---------------- */
#include<stdlib.h>
#include<stdio.h>
#include<math.h>
#include"global.h"
#include"prototipos.h"
/*------------Inicio da Funcao rank2 --------------------------*/
void rank2(int linre[nlmax]){
	int i,modo,li;
	int lista[nlmax];
	double aax;
	double aa[nlmax];
	modo = 1;         // ordem decrescente
	/*---- Normalizando os indices de severidade pelo maior indice ----*/
	//nsobr =10;
	if(pindex == 0)
		pindex = 1;
	for(i=0; i<nsobr; i++){
		if (pia[i]>pindex)
			pindex = pia[i];
	}
	
	for(i=0; i<nsobr; i++){
		lista[i] = i;
		piaux[i] = pia[i]/pindex;
		aa[i] = pia[i]/pindex;
		//printf("\n Apos a normalizacao do piaux[%d]  = %f --- lista[%d] = %d" , i, piaux[i], i, lista[i]);
		// printf("\n Indice ja normalizado aa[%d] em select.c = %f", i, aa[i]);                   // pia[] ja normalizado
	}

	ordena2(nsobr,aa,lista,modo);

	for(i=0; i<nsobr; i++){
		li = lista[i];
		klasif[i] = linre[li];
		pia[i] = piaux[li];
		//printf("\n Indice ja normal e ordenado pia[%d] = %f --- lista[%d] = %d", i,pia[i], i, lista[i]);
	}
	return;
}

/*--------Fim da Funcao rank2---------------*/

/*---------------Inicio da Funcao reordena------------------*/

void reordena(void){
	int i,ii;
	int linre[nlmax];
	
	for(i=0; i<nsobr; i++)
		linre[i] = klasif[i];
	
	rank2(linre);
  /*------- A.) Corte da lista pelo indice de severidade------------*/
	for(i=0; i<nsobr; i++){
		if(pia[i] >= corte)
			ncrit++;                         // corte da lista
	}
  /*---------B.) Corte da lista pelo tamanho--------------*/
	if(ncrit >= ncorte)
		ncrit = ncorte;     // corte pelo tamanho da lista
	printf("\n Lista Original: nsobr= %d  \n ", nsobr);
	printf("\n Lista Reduzida: ncrit= %d  \n ", ncrit);
	//printf("estou em  reodena em Mestre screening");
	return;
	
}
/*-------------Fim da Funcao reordena--------------*/

/*------------Inicio da Funcao ordena--------------*/

/*Algoritmo Heapsort Knuth vol.3
Rotina que obtem os indices correspondentes a ordenacao parcial do vetor lista.
n     - no.de elementos que se deseja ordenar.
k     - no.de elementos dos vetores aa e lista
aa    - vetor real a ordenar (nao e modificado)
lista - vetor inteiro que retorna os indices ordenados do vetor aa
modo  - chave que determina ordem:
      = 0 (crescente)
      = 1 (decrescente)
      (Obs: inicializar o vetor lista com 0 ate n-1.)*/

void ordena(int n,double aa[nlmax],int lista[nlmax],int modo){
	int i,ir,m,nkl,j,l,aux,k;
	double xk;
	k = n;
	if(n != 0){
		l = k/2 + 1; 
		aux = l - 1;
		m = k ;
		nkl = k - n;
		while(l != 0){
			if(l == 1){
				ir=lista[m-1];
				xk=aa[ir];
				lista[m-1]=lista[0];
				m=m-1;
				if(m == nkl || m == 1){
					lista[0]=ir;
					break;
				}
			}
			else{
				l=l-1;
				ir = lista[l - 1];
				xk = aa[ir];
			}
			j=l;
			i=j;
			j=2*j;
			while(j-m <= 0){
				if(j-m < 0){
					if(modo == 0){  // Ordem crescente
						if(aa[lista[j-1]] > aa[lista[j]])
							j++;
						if(xk <= aa[lista[j-1]])
							break;
					}
					else{  // Ordem decrescente
						if(aa[lista[j-1]] < aa[lista[j]])
							j++;
						if(xk >= aa[lista[j-1]])
							break;
					}
				}// fim if(j-m < 0)
				else{
					if(modo == 0){// Ordem crescente
						if(xk <= aa[lista[j-1]])
							break;
					}
					else{  // Ordem decrescente  
						if(xk >= aa[lista[j-1]])
							break;
					}
				}
				lista[i-1] = lista[j-1];
				i = j ;
				j = 2 * j;
			}// fim while(j-m <= 0)
		lista[i-1]=ir;
		} // fim  while(l != 0)
		for(i=k-1;i>=aux;i--){
			ir=lista[k-i-1];
			lista[k-i-1]=lista[i];
			lista[i]=ir;
		}
	}//fim if(n != 0)
	return;
}

/*-----Fim da Funcao heapsort-------------*/

void ordena2(int n,double aa[],int lista[],int modo){
	int i,ir,j,k,xx;
	double xk;
	xk = 0;
	k = 0;
	xx = n;
	for (j = 0; j < n; j++){
		for(i = k; i < xx ;i++){
			if ( aa[i] >= xk){
				xk = aa[i];
				ir = i;
			}
		}
		lista[j] = ir;
		if(ir == k)
			k++;
		if(ir == xx)
			xx = xx - 1;

		aa[ir] = -1;
		xk = 0;
	}
	return;
}

File: estagio/conting/PIPELINE/select.f
c23456781234567812345678123456781234567812345678123456781234567812345678
c   1       2       3       4       5       6       7       8       9
c ----------------------------------------------------------------------
c
c ----------------------------------------------------------------------
c Subrotinas associadas a selecao de contingencias < origem: pl5.f >
c ----------------------------------------------------------------------
      subroutine rank2(linre)
c
      include 'globvar.inc'
      include 'varcont.inc' 
      include 'varsel.inc'
c
      dimension linre(nlmax) 
      dimension lista(nlmax)
      dimension aa(nlmax)

	
      modo=1                 ! ordem decrescente
      do 38 i=1,nsobr
        lista(i)=i
        piaux(i)=pia(i)
        aa(i)=pia(i)
   38 continue
      call ordena( nsobr, nsobr, aa, lista, modo )
      
      do 37 i = 1,nsobr
	li = lista(i)
	lir = linre(li)
	klasif(i) = lir 
   	            
	aax = piaux(li)
	pia(i) = aax

   37 continue
	
      return
      end


c ----------------------------------------------------------------------
c	[ reordena ranking  ]

      subroutine reordena
c
      include 'globvar.inc'
      include 'varcont.inc' 
      include 'varsel.inc'		
c        		
      dimension linre(nlmax)
c
      do i = 1,nsobr
	ii = klasif(i)
 	linre(i) = ii
      enddo

      call rank2(linre)

      
c  A.] Corte da lista pelo indice de severidade
      do i = 1,nsobr
        if( pia(i) .ge. corte ) ncrit = ncrit + 1  ! corte da lista 
      enddo

c  B.] Corte da lista pelo tamanho  

      if( ncrit .ge. ncorte ) ncrit = ncorte       ! corte pelo tamanho da lista
cc      write(*,*)' Lista Original: nrc   = ',nsobr     
cc      write(*,*)' Lista Reduzida: ncrit = ',ncrit     
 
      return
      end                ! Fim da subrotina reordena

c ----------------------------------------------------------------------

      subroutine ordena( n, k, aa, lista, modo )
c
c      Algoritmo Heapsort Knuth Vol. 3

c      Rotina que obtem os indices correspondentes a 
c      ordenacao parcial do vetor lista.
c
c      n      - no. de elementos que se deseja ordenar.
c      k      - no. de elementos dos vetores a e lista
c      aa     - vetor real a ordenar (nao e modificado)
c      lista  - vetor inteiro que retorna os indices 
c               ordenados do  vetor real aa.
c      modo   - chave que determina ordem:
c             = 0 (crescente)
c             = 1 (decrescente)
c
c      (obs: inicializar o vetor lista com 1 ate' n.)
c

      include 'globvar.inc'
      include 'varcont.inc' 
      include 'varsel.inc'

      dimension lista(nlmax)
      dimension aa(nlmax)
  

      if( n .ne. 0 )then
      l = k/2 + 1
      m = k
      nkl = k - n              ! nkl chamava-se nl --> foi substituido


20    if( l .ne. 1 )then 

      l = l - 1
      ir = lista(l)
      xk = aa(ir)
      go to 40
      endif

30       ir = lista(m)
         xk = aa(ir)
         lista(m) = lista(1)
         m = m - 1
         if( m .eq. nkl .or. m .eq. 1 ) then
            lista(1) = ir
            go to 100
         end if
40    j = l
50    i = j
      j = 2*j
      if( j - m ) 60, 70, 80
60       if( modo .eq. 0 ) then
            if(aa(lista(j)).gt.aa(lista(j+1))) j = j + 1
         else
            if(aa(lista(j)).lt.aa(lista(j+1))) j = j + 1
         end if
70          if( modo .eq. 0 ) then
               if(xk.le.aa(lista(j)))go to 80
            else
               if(xk.ge.aa(lista(j)))go to 80
            end if
            lista(i)=lista(j)
            go to 50
80             lista(i)=ir
               go to 20
c
100   continue
      do 90 i = k, (int(k/2) + 1), -1
         ir = lista(k-i+1)
         lista(k-i+1) = lista(i)
         lista(i) = ir
90    continue

	endif

1000  continue
      return
      end

